OsmAnd/src/net/osmand/plus/poi/PoiHelper.java

```diff
@@ -0,0 +1,321 @@
+
+public class PoiHelper {
+
+	private static final Log LOG = PlatformUtil.getLog(PoiHelper.class);
+	private OsmandApplication app;
+	private PoiDbHelper helper;
+	private Map<String, Long> files;
+	private Map<String, List<String>> categories;
+
+	public PoiHelper(OsmandApplication app) {
+		this.app = app;
+		helper = new PoiDbHelper(app);
+	}
+
+	public Map<String, Long> getFiles() {
+		if (files == null) {
+			files = helper.getFiles();
+		}
+		return files;
+	}
+
+	public Map<String, List<String>> getCategories() {
+		if (categories == null) {
+			categories = helper.getCategories();
+		}
+		return categories;
+	}
+
+	public void readPoiTypesFromMap() {
+		LocalIndexHelper localIndexHelper = new LocalIndexHelper(app);
+		List<LocalIndexInfo> localMapsIndexes = localIndexHelper.getLocalFullMaps(new AbstractLoadLocalIndexTask() {
+			@Override
+			public void loadFile(LocalIndexInfo... loaded) {
+			}
+		});
+		Map<String, Long> savedFiles = getFiles();
+		if (savedFiles.size() == localMapsIndexes.size()) {
+			for (LocalIndexInfo info : localMapsIndexes) {
+				File f = new File(info.getPathToData());
+				String name = f.getName();
+				long date = f.lastModified();
+				if (!savedFiles.containsKey(name) || savedFiles.get(name) != date) {
+					initCategoriesFromFiles();
+					replaceSavedFiles(localMapsIndexes);
+					return;
+				}
+			}
+		} else {
+			initCategoriesFromFiles();
+			replaceSavedFiles(localMapsIndexes);
+			return;
+		}
+		readCategoriesFromDb();
+	}
+
+	@SuppressLint("StaticFieldLeak")
+	public void readPoiTypesFromMapAsync() {
+		new AsyncTask<Void, Void, Void>() {
+
+			@Override
+			protected Void doInBackground(Void... voids) {
+				app.getPoiTypes().init();
+				readPoiTypesFromMap();
+				return null;
+			}
+		}.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR);
+	}
+
+	private void readCategoriesFromDb() {
+		for (Map.Entry<String, List<String>> entry : getCategories().entrySet()) {
+			PoiCategory poiCategory = app.getPoiTypes().getPoiCategoryByName(entry.getKey(), true);
+			for (String s : entry.getValue()) {
+				PoiType poiType = new PoiType(MapPoiTypes.getDefault(), poiCategory, null, s);
+				List<String> filters = new ArrayList<>();
+				for (PoiType poi : poiCategory.getPoiTypes()) {
+					filters.add(poi.getKeyName());
+				}
+				if (!filters.contains(s)) {
+					poiCategory.getPoiTypes().add(poiType);
+				}
+			}
+		}
+	}
+
+	private void replaceSavedFiles(List<LocalIndexInfo> localMapsIndexes) {
+		files.clear();
+		helper.deleteFilesTable();
+		for (LocalIndexInfo info : localMapsIndexes) {
+			File f = new File(info.getPathToData());
+			helper.addFile(f);
+			files.put(f.getName(), f.lastModified());
+		}
+		helper.close();
+	}
+
+	private void initCategoriesFromFiles() {
+		app.getPoiTypes().clearCreatedCategories();
+		final BinaryMapIndexReader[] currentFile = app.getResourceManager().getPoiSearchFiles();
+		for (BinaryMapIndexReader r : currentFile) {
+			try {
+				r.initCategories();
```

Comment:
BinaryMapIndexReader shouldn't use at all MapPoiTypes, looks like it was bad design from previous request.